### DISCUSSION TITLE
`assert.Panics` not capturing when a Go routine panics.

### DISCUSSION THREAD

**Discussion by lccmrx** [stretchr/testify#1617] [Q&A] [answered]
I'm writing a test case to eventually capture panics in my app. But eventually, they could come from within the execution of a Go Routine.

I've created a quick test case to prove my point here.

As described, the `Panics` function requires a parameter (`assert.PanicTestFunc`) that is a function signature for a simple function.
```go
func TestGoRoutinePanic(t *testing.T) {
	assert.Panics(t, func() {
		go func() {
			panic("timeout")
		}()
		time.Sleep(5 * time.Second)
	})
}
```
From within this caller, I've spawned a new Go routine that immediately panics, and the result of running this is:

```sh
panic: timeout

goroutine 36 [running]:
test.TestGoRoutinePanic.func1.1()
	/.../Desktop/main_test.go:13 +0x2c
created by test.TestGoRoutinePanic.func1 in goroutine 35
	/.../Desktop/main_test.go:12 +0x24
FAIL	test	0.763s
FAIL
```

So the assert failed to capture a panic that happened from within the passed function.
removing the `go` token and synchronising the execution that panics, the result is:

```go
func TestGoRoutinePanic(t *testing.T) {
	assert.Panics(t, func() {
		func() {
			panic("timeout")
		}()
		time.Sleep(5 * time.Second)
	})
}
```
```sh
ok  	test	1.004s
```

Could someone tell me why this is?

**Comment by Antonboom** [comment, marked as answer]
`panic` and `recover` works in context of current goroutine only

- https://go.dev/blog/defer-panic-and-recover
- https://go.dev/ref/spec#Handling_panics
- https://github.com/golang/go/issues/55038

**Reply by lccmrx** [reply]
Fair enough!

The same goes for fatal panics?
Cause the test I'm trying to make triggers a fatal panic.
